### PR TITLE
SL-1470: Implement basic tabs in UIKit

### PR DESCRIPTION
**Pull request by XVincentX**
This PR implements a basic Tabs that'll be used to complete the request maker component.

Given this is nothing more than a styled wrapper for another library, I'm not completely sure what I should be testing here, so any suggestion here is welcomed!

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -54,6 +54,7 @@
     "react-input-autosize": "2.2.x",
     "react-select": "2.1.x",
     "react-simple-code-editor": "0.8.x",
+    "react-tabs": "^3.0.0",
```
Comment (P0lip) [inline package.json]: Let's define an exact version here (without `^`)

File: src/Tabs/Tab.tsx [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import { ClassNames, css } from '@emotion/core';
```
Comment (P0lip) [inline src/Tabs/Tab.tsx]: You are missing `@jsx` pragma.

Comment (P0lip) [reply src/Tabs/Tab.tsx]:
```suggestion
/* @jsx jsx */
import { ClassNames, css, jsx } from '@emotion/core';
```

File: src/Tabs/TabPanel.tsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { ClassNames } from '@emotion/core';
```
Comment (P0lip) [inline src/Tabs/TabPanel.tsx]:
```suggestion
/* @jsx jsx */
import { ClassNames, jsx } from '@emotion/core';
```

File: src/Tabs/Tab.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        </Tab>
+      )}
+    </ClassNames>
+  );
+};
+
+StyledTab.tabsRole = 'Tab';
+
+const tabStyles = (tabTheme: { fg: string; bg: string }) => {
+  return {
```
Comment (P0lip) [inline src/Tabs/Tab.tsx]: Could we use theme hook here?
I know it doesn't really matter, but we had such a convention in all components.

File: src/Tabs/Tab.tsx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    tabStyle: css({
```
Comment (P0lip) [inline src/Tabs/Tab.tsx]: You don't need `css` here.
A plain object literal is more than enough :)

File: package.json [hunk 6]
```diff
@@ -68,21 +69,22 @@
...
     "@types/react-window": "1.1.x",
+    "@types/styled-system": "3.0.x",
     "react": "^16.7.0-alpha.2",
     "react-dom": "^16.7.0-alpha.2",
     "storybook-chromatic": "1.2.x",
     "typescript": "3.2.2"
   },
   "lint-staged": {
-    "xxx": [
+    "*.{ts,tsx}": [
```
Comment (P0lip) [inline package.json]: :+1:

File: src/Tabs/Tab.tsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      cursor: 'pointer',
+      backgroundColor: tabTheme.bg,
+      color: tabTheme.fg,
+
+      '&:focus': {
+        boxShadow: '0 0 5px hsl(208, 99%, 50%)',
+        borderColor: 'hsl(208, 99%, 50%)',
+        outline: 'none',
+
+        '&:after': {
```
Comment (P0lip) [inline src/Tabs/Tab.tsx]: `::after` is [preferred](https://developer.mozilla.org/en-US/docs/Web/CSS/::after) over `:after`

File: src/Tabs/Tab.tsx [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+      position: 'relative',
+      listStyle: 'none',
+      padding: '6px 12px',
+      cursor: 'pointer',
+      backgroundColor: tabTheme.bg,
+      color: tabTheme.fg,
+
+      '&:focus': {
+        boxShadow: '0 0 5px hsl(208, 99%, 50%)',
+        borderColor: 'hsl(208, 99%, 50%)',
```
Comment (P0lip) [inline src/Tabs/Tab.tsx]: We are missing a themed color here.
image

You can switch between themes using our theme addon (click on theme tab)
image

File: src/Tabs/TabList.tsx [hunk 9]
```diff
@@ -0,0 +1,21 @@
+/* @jsx jsx */
+
+import { css, jsx } from '@emotion/core';
+import * as React from 'react';
+import { TabList } from 'react-tabs';
```
Comment (P0lip) [inline src/Tabs/TabList.tsx]:
```suggestion
import { TabList, TabListProps } from 'react-tabs';
```

File: src/Tabs/TabList.tsx [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+const StyledTabList: React.FunctionComponent & { tabsRole: string } = props => (
```
Comment (P0lip) [inline src/Tabs/TabList.tsx]: `react-tabs` seems to export props, so let's make use of them.
```suggestion
const StyledTabList: React.FunctionComponent<TabListProps> & { tabsRole: string } = props => (
```

File: src/Tabs/Tab.tsx [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        outline: 'none',
+
+        '&:after': {
+          content: '""',
+          position: 'absolute',
+          height: '5px',
+          left: '-4px',
+          right: '-4px',
+          bottom: '-5px',
+          background: '#fff',
```
Comment (P0lip) [inline src/Tabs/Tab.tsx]: Theme

File: src/Tabs/Tab.tsx [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+        },
+      },
+    }),
+
+    selectedTabStyle: css({
+      background: '#fff',
```
Comment (P0lip) [inline src/Tabs/Tab.tsx]: Ditto

File: src/Tabs/Tab.tsx [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+      borderColor: '#aaa',
```
Comment (P0lip) [inline src/Tabs/Tab.tsx]: Ditto

File: src/Tabs/Tab.tsx [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+      color: 'black',
+      borderRadius: '5px 5px 0 0',
+    }),
+
+    disabledTabStyle: css({
+      color: 'GrayText',
```
Comment (P0lip) [inline src/Tabs/Tab.tsx]: Ditto

File: src/Tabs/TabList.tsx [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+  <TabList {...props} css={tabListStyle}>
+    {props.children}
+  </TabList>
+);
+
+StyledTabList.tabsRole = 'TabList';
+
+const tabListStyle = css({
+  borderBottom: '1px solid #aaa',
```
Comment (P0lip) [inline src/Tabs/TabList.tsx]: Ditto

File: src/Tabs/Tab.tsx (hunk 5 above)
Comment (P0lip) [reply src/Tabs/Tab.tsx]: If TS complains, make sure to annotate return.
```tsx
const tabStyles = (tabTheme: { fg: string; bg: string }): Dictionary<IBoxCSS> => {
```
should do the trick for you

File: src/Tabs/TabPanel.tsx (hunk 3 above)
Comment (XVincentX) [reply src/Tabs/TabPanel.tsx]: Is this required even though I'm not using any special feature of emotion's `jsx?`

File: src/Tabs/Tab.tsx (hunk 4 above)
Comment (XVincentX) [reply src/Tabs/Tab.tsx]: Conventions are important! I'll fix!

File: src/Tabs/TabPanel.tsx (hunk 3 above)
Comment (P0lip) [reply src/Tabs/TabPanel.tsx]: In general, it's not required in your particular case right now, but I'd add it, as we do that in the majority of our components.
Anyway, I won't complain if you leave it as is.
I believe emotion's jsx is needed only if you do the following `<span css={styles}>a</span>`

File: src/Tabs/Tab.tsx (hunk 7 above)
Comment (XVincentX) [reply src/Tabs/Tab.tsx]: Ah nice, I didn't know it!

File: src/Tabs/Tab.tsx (hunk 4 above)
Comment (XVincentX) [reply src/Tabs/Tab.tsx]: Done in 8bdac75

File: src/Tabs/Tab.tsx (hunk 5 above)
Comment (XVincentX) [reply src/Tabs/Tab.tsx]: Done!

**Comment by XVincentX** [comment]
@P0lip Can you have another look? Hopefully I haven't forgot anything.

File: src/Tabs/Tab.tsx (hunk 8 above)
Comment (P0lip) [reply src/Tabs/Tab.tsx]: @XVincentX bump ^

File: src/Tabs/Tab.tsx [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+      bottom: '-1px',
+      position: 'relative',
+      listStyle: 'none',
+      padding: '6px 12px',
+      cursor: 'pointer',
+      backgroundColor: theme.tabs.bg,
+      color: theme.tabs.fg,
+
+      '&:focus': {
+        boxShadow: '0 0 5px hsl(208, 99%, 50%)',
```
Comment (P0lip) [inline src/Tabs/Tab.tsx]: let's add a theme color for this shadow as well (follow `blockQuote` naming)

File: src/theme/light.ts [hunk 17]
```diff
@@ -116,6 +116,13 @@ export const lightTheme: ITheme = {
     bg: '#fff',
   },
 
+  tabs: {
+    fg: '#222',
+    bg: '#fff',
+    border: '#aaa',
+    disabledColor: 'GrayText',
```
Comment (P0lip) [inline src/theme/light.ts]: `disabledFg`

**Comment by XVincentX** [comment]
I admit I have no taste at all for styling, that's why I was asking for a style guide yesterday :)

File: src/Tabs/Tab.tsx (hunk 8 above)
Comment (P0lip) [reply src/Tabs/Tab.tsx]: Apparently, it's some blue color which looks fine for both light and dark theme, so let's leave it for now.

**Pull request merged**